blog/2022/04/simple-chains.md

```diff
@@ -0,0 +1,493 @@
+  W = ...
+  b = ...
+  r = W*x .+ b
+  y = f.σ.(r)
+  function FastDense_adjoint(v)
+    σbar = ForwardDiff.derivative.(f.σ,r)
+    zbar = v .* σbar
+    Wbar = zbar * x'
+    bbar = zbar
+    xbar = W' * zbar
+    nothing,xbar,vcat(vec(Wbar),bbar)
+  end
+  y,FastDense_adjoint
+end
+```
+
+for `sigma.(W*x .+ b)` to calculate `J'v`, you can greatly optimize this if you know that the backwards pass will immediately preceed the forward pass. Specifically, there's no need to generate closures to store values since there no indeterminate future where the gradient might be needed, instead you can immediately proceed to calculate it. And, if you're only applying it to a vector of known size `v`, then this operation can be done without allocating by mutating a cache vector. Lastly, if we know we're only going to use the derivative w.r.t. `x` (`xbar`), then we can eliminate many calculations. Look at the simplified version:
+
+```julia
+r = W*x .+ b
+y = σ.(r)
+σbar = derivative.(σ,r)
+zbar = v .* σbar
+Wbar = zbar * x'
+bbar = zbar
+xbar = W' * zbar   
+```
+
+and now cached:
+
+```julia
+mul!(cache,W,x)
+cache .= σ.(cache .+ b)
+cache .= derivative.(σ,cache)
+cache .= v .* cache
+mul!(xbar,W',cache)
+```
+
+or in other words, we can write this as a mutating operation with a single cache vector: `vjp!(xbar,W,b,σ,v,cache)`. All of the overheads of any automatic differentiation or mutations? Gone.
+
+Of course, building this up for anything other than the simplest case takes a much larger effort. In comes SimpleChains.jl.
+
+## SimpleChains.jl: An Optimized Machine Learning Library for SciML Use Cases
+
+[SimpleChains.jl](https://github.com/PumasAI/SimpleChains.jl) is the solution to this problem. SimpleChains.jl is a small machine learning framework optimized for quickly fitting small models on the CPU. Early development favored a design that would:
+
+1. Allow us to achieve good performance, ideally approaching the CPU's potential peak FLOPs.
+2. Focus on small size meant we could largely forgo large kernel optimizations (such as cache tiling) in the early stages of development.
+3. Have an API where vectors of parameters (and their gradients) are first class, rather than having parameters live with the layers, to make it easier to work with various optimizers or solvers that expect contiguous vectors (such as BFGS).
+4. Be written in "pure Julia" for ease of development and optimization; while making heavy use of [LoopVectorization.jl](https://github.com/JuliaSIMD/LoopVectorization.jl/), SimpleChains.jl does not rely on any BLAS or NN libraries. It is a long term aim to extend this loop-compiler approach to optimization to also producing pullbacks automatically, without requiring them to be handwritten. However, the compiler-focused approach is already levered for ease of implementation: while we still have to hand-write gradients, we do not need to hand-optimize them.
+
+## SimpleChains.jl in Action: 10x-ing PyTorch in Small Examples
```

Comment:
Well, the benchmarks here are not quite 10x. Some of them a fair bit less than 10x.